### ISSUE TITLE
Question: plugins and TCP

### ISSUE THREAD

**Issue by nubunto** [vernemq/vernemq#1130]
Not really an issue, but a question.

Today, we have a running EMQ X 2.2 broker in production. However, some of the features of version 3 are present in VerneMQ, and the developer experience with the EMQ X broker is... not that great.

Today, we have built a custom plugin that checks our OAuth tokens and publishes message to AWS SNS. We want to do the same with VerneMQ.

Currently, we expose the websocket endpoint for the outside world so our clients can connect directly to our broker, and we keep the TCP endpoint inside our VPC so that we can integrate with the existing infrastructure and skip this validation.

However, with EMQ X we are able to get the Websocket headers and allow internal systems to do whatever they like with the broker. I don't see a clear way to make this distinction with the VerneMQ plugin that I'm currently writing. The plugin gets invoked for TCP and Websocket connections, which is great, but we can't really make this distinction.

So, in short: is there a way we can activate a plugin for a given protocol or something similar? Perhaps we are missing something or maybe there's another simpler way to achieve what we're looking for.

**Comment by nubunto** [comment]
FWIW, I was able to make this distinction in the `mountpoint`.

So, in my `vernemq.conf`:
```
listener.tcp.default.mountpoint = internal # TCP connections only happen inside the VPC
listener.ws.default.mountpoint = websocket
listener.wss.default.mountpoint = websocket
```

Then, in my plugin, I can happily let `internal` mountpoints bypass the `auth_on_register`:
```
def auth_on_register(_peer, {'internal', clientid}, username, _password, _clean_session?) do
      IO.puts("*** auth_on_register called internally #{clientid}")
      {:ok, []}
    end
    def auth_on_register(peer, {'websocket', clientid}, _username, _password, _clean_session?) do
      IO.puts("*** auth_on_register called via websocket")
      {:ok, []}
    end
    def auth_on_register(_peer, _info, _username, _password, _clean_session?) do
      :error
    end
```

**Comment by larshesel** [comment]
Hi, great you found out how to do it. Using mountpoints is indeed the recommended way the moment. We have a plan to make it possible to assign plugins to specific listeners so for example an external listener uses a db plugin for handling authentication while an internal listener uses file based authentication. That would give a bit more flexibility.
